### PR TITLE
[HW][Seq] Added traits to identify clocked and resettable elements

### PR DESCRIPTION
**Pull request by nandor**
Added `Clocked` and `Resettable` traits to the `hw` dialect. Some `seq` ops now conform to these traits.

The goal of `Clocked` and `Resettable` is to generically identify the special wires of sequential elements in analyses without needing to special-case all register types.

### CODE REVIEW

**Comment by uenoku** [comment]
I think the trait makes sense to me, but I feel Seq is better to place than HW for this abstraction. Do you have reason to have this in HW?

**Comment by nandor** [comment]
Yes - we have assert and printf ops that are also users of clocks and I think those should be marked as well. I think HW is a good common place to put these traits in for now.

File: include/circt/Dialect/HW/HWOpInterfaces.td [hunk 1]
```diff
@@ -157,4 +157,23 @@ def HWInstanceLike : OpInterface<"HWInstanceLike"> {
   ];
 }
 
+def Clocked : OpInterface<"Clocked"> {
+  let cppNamespace = "circt::seq";
```
Comment (uenoku) [inline include/circt/Dialect/HW/HWOpInterfaces.td:161]: Does this need to be `hw`?

**Comment by teqdruid** [comment]
> Yes - we have assert and printf ops that are also users of clocks and I think those should be marked as well. I think HW is a good common place to put these traits in for now.

Is there any reason that those `sv` ops can't include traits from the `seq` dialect? The `seq` dialect shouldn't need anything from the `sv` dialect, so I don't think there'd be any dependency cycle. The seq-to-sv lowering pass is a separate thing (or should be).

**Review by teqdruid — APPROVED** [review]
I like this! I would prefer it be in the `seq` dialect. Based on the `cppNamespace` thing, looks like that's what you did originally.

I'm pretty sure [this is an unnecessary dep](https://github.com/llvm/circt/blob/257e50d0321640f6dc06a5c34a14606c0a4849cc/lib/Dialect/Seq/CMakeLists.txt#L30). (Which is my bad.)

**Comment by nandor** [comment]
You're right - I think hw should not know about clocking and it makes sense to add a dependency between sv and seq.

File: include/circt/Dialect/HW/HWOpInterfaces.td (hunk 1 above)
Comment (nandor) [reply include/circt/Dialect/HW/HWOpInterfaces.td:161]: Nope, we're back to sv.

**Pull request merged**